core/server/data/migrations/versions/4.12/02-fix-member-statuses.js

```diff
@@ -5,10 +5,21 @@ module.exports = createTransactionalMigration(
     async function up(knex) {
         logging.info('Updating members with no products to "free"');
 
-        await knex('members')
+        const freeMembers = await knex('members')
+            .select('members.id')
             .leftJoin('members_products', 'members.id', 'members_products.member_id')
-            .update('status', 'free')
             .whereNull('members_products.member_id');
+
+        if (freeMembers.length === 0) {
+            logging.info('No free members found');
+            return;
+        }
+
+        const freeMembersIds = freeMembers.map(row => row.id);
+
+        await knex('members')
```

Comment:
Think it's best to keep the "chunking" [pattern](https://github.com/TryGhost/Ghost/blob/main/core/server/data/migrations/versions/4.0/03-populate-status-column-for-members.js#L65-L69) here even though the SQLite parameter limit is way bigger now.

Reply:
Hmm, so the new limit is 32766 - this seems like more than enough to me, unless we have sites out there that have 32k cancelled subscriptions!

I'll add the chunking anyway, but would be good to have a conversation about when that's considered premature optimisation or not

Reply:
I also think it's a long stretch to have this many cancelled subscriptions out there. The reason why I raised the topic is because we are meant to write code "as if there is a 100k members" in the system from perf perspective. With a big audience of ~100k the 32k limit starts becoming possible? Maybe I'm being too paranoid about having bullet proof code, or maybe we should make this kind of update helper and forget about it :sweat_smile:

Reply:
2 things 

1. the new version of sqlite3 that has the new limit isn't yet widely accessible or available in Node. We're trying to take over the node package to fix this, but it's taking time.
2. We are meant to write code as if there is 1million members in the system, as documented here: https://www.notion.so/ghost/Database-Migrations-Seeding-08a408efbd9140e691f46cb903592fd7

Reply:
Oh, my bad - I thought since we'd bumped the `sqlite3` to 5.x we'd be using SQLite3 3.34 :grimacing:

So we should always assume that we're working with SQLite3 < 3.32, right?

Reply:
We're currently shipping with `sqlite3` 5.0.2, which bundles SQLite 3.34 and the limit was bumped to 32766 in 3.32.0.

However, I'm not sure that the limit applies for `where in (...`? Still, I wouldn't fancy sending a LOT of un-chunked data in 1 query so I think this looks good.

Reply:
The limit applies to host parameters, which knex uses under the hood like this:

```sql
SELECT column FROM table WHERE id IN (?, ?, ?) VALUES (1, 2, 3)
```

The limit is essentially on the number of `?` in the statement

ref https://sqlite.org/limits.html (point 9.)

Reply:
TIL - ok!